### PR TITLE
Integrate the work done by @jdeneef back to this repo

### PR DESCRIPTION
**Pull request by tomer-w**
This will allow Raspberry Pi 5 support and a few more advanced tweaks.
Based on the discussion here: https://github.com/thecode/ha-rpi_gpio/issues/278

### CODE REVIEW

**Comment by tomer-w** [comment]
@jdeneef, @raboof, FYI.

**Comment by tomer-w** [comment]
Closed by mistake.

**Comment by thecode** [comment]
@tomer-w when we discussed about making a PR to update this repo my intention was not just to merge another repo into this one, otherwise I could do it without a PR.

Please remove anything which is not necessary from this PR, docs, license, json file updates should not be in this PR, focus on only code changes if other changes are needed please move them to a different PR.
In addition if you want this PR to reviewed faster try to split it into multiple PRs, I can create a branch so we merge PRs into that branch and merge it to main when ready, otherwise the review process will take time.

**Comment by tomer-w** [comment]
Sure, feel free to create a branch. It will make it easier.
This PR is not only merging different repo, it is aligning it back with the original yaml format so it is backwards compatible.
Most of the code is coming from the source repo so there is enough data that it is actually working. I think most of the focus should be on the docs, json, etcs. 
Anyhow I published it that we will have something to work with, branch will be the best solution.

**Comment by thecode** [comment]
> Sure, feel free to create a branch. It will make it easier. 

https://github.com/thecode/ha-rpi_gpio/tree/gpiod-rpi5

**Comment by tomer-w** [comment]
PR opened. Can you give me write access to that branch?

**Comment by tomer-w** [comment]
I moved this PR to the new branch. Please approve and give me write access so we can iterate over it.

**Comment by thecode** [comment]
I think my comment was misunderstood. I asked to split this PR into several PRs made to the new branch, not that you merge content into this branch without review.

**Comment by tomer-w** [comment]
I removed from the PR all the auxiliary changes. We now have only the actual code, the documentation which states the new parameters and a couple more super minor changes. It is only 11 files, 7 of them are code files. Hope this is small enough we can review.

**Comment by thecode** [comment]
Please split the docs into a different PR, I will start reviewing code changes

File: custom_components/rpi_gpio/const.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
+DOMAIN = "gpiod"
```
Comment (thecode) [inline custom_components/rpi_gpio/const.py:2]:
```suggestion
DOMAIN = "rpi_gpiod"
```
I know this is now supporting other platforms, but the name needs to match the intergration

File: custom_components/rpi_gpio/cover.py [hunk 2]
```diff
@@ -50,88 +53,100 @@
...
 
     def close_cover(self, **kwargs):
-        """Close the cover."""
-        if not self.is_closed:
-            self._trigger()
+        if self.is_closed:
+            return
+        self._hub.turn_on(self._relay_port)
+        self.is_closing = True
+        # self.is_closed = None
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:121]:
```suggestion
```
Please remove all dead code

File: custom_components/rpi_gpio/hub.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        if not self._online:
+            return
+
+        # setup lines
+        self.update_lines()
+
+        # read initial states for sensors
+        # for port in self._config:
+            # self._entities[port].update()
+
```
Comment (thecode) [inline custom_components/rpi_gpio/hub.py:103]: Dead code

File: custom_components/rpi_gpio/manifest.json [hunk 4]
```diff
@@ -1,10 +1,11 @@
 {
   "domain": "rpi_gpio",
   "name": "Raspberry Pi GPIO",
-  "codeowners": ["@thecode"],
+  "codeowners": [ "@thecode", "@jdeneef", "@tomer-w" ],
```
Comment (thecode) [inline custom_components/rpi_gpio/manifest.json:4]: This will be done in a separate PR

File: custom_components/rpi_gpio/switch.py [hunk 5]
```diff
@@ -1,160 +1,116 @@
...
-        self._state = True if state.state == STATE_ON else False
-        if self._state:
-            await self.async_turn_on()
-        else:
-            await self.async_turn_off()
\ No newline at end of file
+    def update(self):
+        self.is_on = self._hub.update(self._port)
+        self.schedule_update_ha_state(False)
+
```
Comment (thecode) [inline custom_components/rpi_gpio/switch.py:116]:
```suggestion
```

**Review by thecode — CHANGES_REQUESTED** [review]
- Remove docs from this PR
- Remove all commented out code
- make sure domain name doesn't change-
-  When hub is offline we should handle this better, not by just logging an error

File: custom_components/rpi_gpio/const.py (hunk 1 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/const.py]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 2 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:121]: fixed.

File: custom_components/rpi_gpio/hub.py (hunk 3 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/hub.py]: fixed

File: custom_components/rpi_gpio/manifest.json (hunk 4 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/manifest.json]: ok

File: custom_components/rpi_gpio/switch.py (hunk 5 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/switch.py]: fixed

**Comment by tomer-w** [comment]
I fixed all the CR comments, re-tested the integration and removed the minor non code changes other than the documentation itself as I dont want to lose the changes. Can we finish the CR as is? I can spend time about splitting the single MD file but it is easier to check it in into the branch and then iterate on it.

File: README.md [hunk 6]
```diff
@@ -183,4 +208,12 @@ switch:
...
-```
+
+
+# Add Debug info and issue reporting
+*Before* reporting issues please add this to your `configuration.yaml` `logger` section, check logs and report issue adding logging.
+```yaml
+logger:
+  default: info
+  logs:
+    custom_components.rpi_gpio: debug
```
Comment (thecode) [inline README.md:219]: Please link to https://www.home-assistant.io/docs/configuration/troubleshooting/#enabling-debug-logging instead. There is no need to use YAML anymore for debug and the UI functionality automatically download the log upon disable.

File: hacs.json [hunk 7]
```diff
@@ -1,4 +1,5 @@
 {
   "name": "Raspberry Pi GPIO",
-  "homeassistant": "2022.7.0"
+  "homeassistant": "2022.7.0",
+  "render_readme": true
```
Comment (thecode) [inline hacs.json:4]:
```suggestion
  "homeassistant": "2022.7.0"
```
`render_readme` is no longer supported by `HACS`

**Review by thecode — CHANGES_REQUESTED** [review]
Reviewed only the docs for now

File: README.md (hunk 6 above)
Comment (tomer-w) [reply README.md:219]: done

File: hacs.json (hunk 7 above)
Comment (tomer-w) [reply hacs.json:4]: done, but dont you want to bump up the version?

**Comment by tomer-w** [comment]
Fixed all current doc comments. My only question was regarding the versioning which I assumed we should bump.

File: hacs.json (hunk 7 above)
Comment (thecode) [reply hacs.json]: We will bump it before the release

**Comment by jdeneef** [comment]
@tomer-w quick check on your changes shows that you are removing all config validations and parameters names reverting it back to rpi-gpio terminology? I'm a bit puzzled why you would do that? I made those changes on purpose to get closer to the actual used libgpiod library for anybody who really needs to understand what the parameter actually does, not having to reverse engineer the code to find out what parameters is actually set?

**Comment by tomer-w** [comment]
I fully understand your motives and it makes sense from that perspective but as the rpi-gpio integration is heavily used, I didn't want to break things for current users or stress them with documentation which says that those old configuration can be deprecated.
At least this was my thinking, but I'm open to any feedback.

**Comment by jdeneef** [comment]
ok, clear, thanx for explaining! I already had the plan to simply keep both since I did find a nice way without code duplication in the config validation, just hadn't updated docs just yet. So you could mention both new and old config options in one row, keep the config validation as is and remove all 'backwards compatibility' stuff, that could be a fine option in the middle? BTW I did not know it was still used? It stopped working for me like 8 months ago when my distro moved to 6..1.69. So fair enough to make transfer as simple as possible.

**Comment by tomer-w** [comment]
Sound good. I'm waiting for @thecode to merge it into the branch. Than we can work on it together to make sure it is backwards compatible and still can be used with the native naming you used.

**Comment by tomer-w** [comment]
And yes, the integration still works as far as you didn't move to RP5.

File: custom_components/rpi_gpio/binary_sensor.py [hunk 8]
```diff
@@ -1,141 +1,89 @@
...
 
-    def __init__(self, name, port, pull_mode, bouncetime, invert_logic, unique_id=None):
-        """Initialize the RPi binary sensor."""
-        self._attr_name = name or DEVICE_DEFAULT_NAME
-        self._attr_unique_id = unique_id
-        self._attr_should_poll = False
+    def __init__(self, hub, name, port, unique_id, active_low, bias, debounce):
+        _LOGGER.debug(f"GPIODBinarySensor init: {port} - {name} - {unique_id}")
+        self._hub = hub
+        self.name = name
```
Comment (thecode) [inline custom_components/rpi_gpio/binary_sensor.py:75]:
```suggestion
        self._attr_name = name
```

File: custom_components/rpi_gpio/cover.py [hunk 9]
```diff
@@ -50,88 +53,100 @@
...
+    async_add_entities(covers)
+
+class GPIODCover(CoverEntity):
+    should_poll = False
+
+    def __init__(self, hub, name, relay_port, relay_time, relay_active_low, relay_bias, relay_drive,
+                 state_port, state_bias, state_active_low, unique_id):
+        _LOGGER.debug(f"GPIODCover init: {relay_port}:{state_port} - {name} - {unique_id} - {relay_time}")
+        self._hub = hub
+        self.name = name
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:98]:
```suggestion
        self._attr_name = name
```

File: custom_components/rpi_gpio/switch.py [hunk 10]
```diff
@@ -1,160 +1,114 @@
...
+    async_add_entities(switches)
+
+
+class GPIODSwitch(SwitchEntity, RestoreEntity):
+    should_poll = False
+
+    def __init__(self, hub, name, port, unique_id, active_low, bias, drive, persistent):
+        _LOGGER.debug(f"GPIODSwitch init: {port} - {name} - {unique_id} - active_low: {active_low} - bias: {bias} - drive: {drive}")
+        self._hub = hub
+        self.name = name
```
Comment (thecode) [inline custom_components/rpi_gpio/switch.py:81]:
```suggestion
        self._attr_name = name
```

File: custom_components/rpi_gpio/cover.py [hunk 11]
```diff
@@ -50,88 +53,100 @@
...
     def update(self):
-        """Update the state of the cover."""
-        self._state = read_input(self._state_pin)
-
-    @property
-    def is_closed(self):
-        """Return true if cover is closed."""
-        return self._state != self._invert_state
+        self.is_closed = self._hub.update(self._state_port)
+        self.schedule_update_ha_state(False)
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:114]:
```suggestion
```
This should cause update to happen twice, please verify if it is really needed

File: custom_components/rpi_gpio/binary_sensor.py [hunk 12]
```diff
@@ -1,141 +1,89 @@
...
+    async def async_added_to_hass(self) -> None:
+        await super().async_added_to_hass()
+        self._hub.add_sensor(self, self._port, self._active_low, self._bias, self._debounce)
+        self.async_write_ha_state()
 
     def update(self):
-        """Update the GPIO state."""
-        self._state = read_input(self._port)
+        self.is_on = self._hub.update(self._port)
+        self.schedule_update_ha_state(False)
```
Comment (thecode) [inline custom_components/rpi_gpio/binary_sensor.py:89]:
```suggestion
```
This should cause update to happen twice, please verify if it is really needed

File: custom_components/rpi_gpio/switch.py [hunk 13]
```diff
@@ -1,160 +1,114 @@
...
-            return
-        self._state = True if state.state == STATE_ON else False
-        if self._state:
-            await self.async_turn_on()
-        else:
-            await self.async_turn_off()
\ No newline at end of file
+    def update(self):
+        self.is_on = self._hub.update(self._port)
+        self.schedule_update_ha_state(False)
```
Comment (thecode) [inline custom_components/rpi_gpio/switch.py:114]:
```suggestion
```
This should cause update to happen twice, please verify if it is really needed

File: custom_components/rpi_gpio/binary_sensor.py [hunk 14]
```diff
@@ -1,141 +1,89 @@
...
-class RPiGPIOBinarySensor(BinarySensorEntity):
-    """Represent a binary sensor that uses Raspberry Pi GPIO."""
 
-    async def async_read_gpio(self):
-        """Read state from GPIO."""
-        await asyncio.sleep(float(self._bouncetime) / 1000)
-        self._state = await self.hass.async_add_executor_job(read_input, self._port)
-        self.async_write_ha_state()
+class GPIODBinarySensor(BinarySensorEntity):
+    should_poll = False
```
Comment (thecode) [inline custom_components/rpi_gpio/binary_sensor.py:70]:
```suggestion
    _attr_should_poll = False
```

File: custom_components/rpi_gpio/cover.py [hunk 15]
```diff
@@ -50,88 +53,100 @@
...
-        self._attr_unique_id = unique_id
-        self._state = False
-        self._relay_pin = relay_pin
-        self._state_pin = state_pin
-        self._state_pull_mode = state_pull_mode
+
+    async_add_entities(covers)
+
+class GPIODCover(CoverEntity):
+    should_poll = False
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:92]:
```suggestion
    _attr_should_poll = False
```

File: custom_components/rpi_gpio/hub.py [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+}
+DRIVE = { 
+    "OPEN_DRAIN": Drive.OPEN_DRAIN, 
+    "OPEN_SOURCE": Drive.OPEN_SOURCE, 
+    "PUSH_PULL": Drive.PUSH_PULL, 
+} 
+
+class Hub:
+
+    manufacturer = "ha_gpiod"
```
Comment (thecode) [inline custom_components/rpi_gpio/hub.py:33]:
```suggestion
```

File: custom_components/rpi_gpio/switch.py [hunk 17]
```diff
@@ -1,160 +1,114 @@
...
+                switch.get(CONF_DRIVE),
+                switch[CONF_PERSISTENT]
+            )
+        )
+
+    async_add_entities(switches)
+
+
+class GPIODSwitch(SwitchEntity, RestoreEntity):
+    should_poll = False
```
Comment (thecode) [inline custom_components/rpi_gpio/switch.py:76]:
```suggestion
    _attr_should_poll = False
```

File: custom_components/rpi_gpio/switch.py (hunk 10 above)
Comment (thecode) [inline custom_components/rpi_gpio/switch.py:81]:
```suggestion
        self._attr_name = name
```

File: custom_components/rpi_gpio/binary_sensor.py [hunk 18]
```diff
@@ -1,141 +1,89 @@
...
-    def __init__(self, name, port, pull_mode, bouncetime, invert_logic, unique_id=None):
-        """Initialize the RPi binary sensor."""
-        self._attr_name = name or DEVICE_DEFAULT_NAME
-        self._attr_unique_id = unique_id
-        self._attr_should_poll = False
+    def __init__(self, hub, name, port, unique_id, active_low, bias, debounce):
+        _LOGGER.debug(f"GPIODBinarySensor init: {port} - {name} - {unique_id}")
+        self._hub = hub
+        self.name = name
+        self.unique_id = unique_id
```
Comment (thecode) [inline custom_components/rpi_gpio/binary_sensor.py:76]:
```suggestion
        self._attr_unique_id = unique_id
```

File: custom_components/rpi_gpio/cover.py [hunk 19]
```diff
@@ -50,88 +53,100 @@
...
-        setup_output(self._relay_pin)
-        setup_input(self._state_pin, self._state_pull_mode)
-        write_output(self._relay_pin, 0 if self._invert_relay else 1)
+        self._relay_active_low = relay_active_low
+        self._relay_bias = relay_bias
+        self._relay_drive = relay_drive
+        self._state_port = state_port
+        self._state_bias = state_bias
+        self._start_active_low = state_active_low
+        self.unique_id = unique_id
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:107]:
```suggestion
        self._attr_unique_id = unique_id
```

File: custom_components/rpi_gpio/cover.py [hunk 20]
```diff
@@ -50,88 +53,100 @@
...
 
     def update(self):
-        """Update the state of the cover."""
-        self._state = read_input(self._state_pin)
-
-    @property
-    def is_closed(self):
-        """Return true if cover is closed."""
-        return self._state != self._invert_state
+        self.is_closed = self._hub.update(self._state_port)
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:113]:
```suggestion
        self._attr_is_closed = self._hub.update(self._state_port)
```

File: custom_components/rpi_gpio/cover.py [hunk 21]
```diff
@@ -50,88 +53,100 @@
...
+        self._hub.turn_on(self._relay_port)
+        self.is_closing = True
+        self.schedule_update_ha_state(False)
         sleep(self._relay_time)
-        write_output(self._relay_pin, 0 if self._invert_relay else 1)
+        if not self.is_closing:
+            # closing stopped
+            return
+        self._hub.turn_off(self._relay_port)
+        self.is_closing = False
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:127]:
```suggestion
        self._attr_is_closing = False
```

File: custom_components/rpi_gpio/cover.py [hunk 22]
```diff
@@ -50,88 +53,100 @@
...
+        self.schedule_update_ha_state(False)
 
-    def _trigger(self):
-        """Trigger the cover."""
-        write_output(self._relay_pin, 1 if self._invert_relay else 0)
+    def close_cover(self, **kwargs):
+        if self.is_closed:
+            return
+        self._hub.turn_on(self._relay_port)
+        self.is_closing = True
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:120]:
```suggestion
        self._attr_is_closing = True
```

File: custom_components/rpi_gpio/cover.py [hunk 23]
```diff
@@ -50,88 +53,100 @@
...
+        self.update()
 
-    def close_cover(self, **kwargs):
-        """Close the cover."""
+    def open_cover(self, **kwargs):
         if not self.is_closed:
-            self._trigger()
+            return
+        self._hub.turn_on(self._relay_port)
+        self.is_opening = True
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:134]:
```suggestion
        self._attr_is_opening = True
```

**Review by thecode — CHANGES_REQUESTED** [review]
Please go over all of the attributes and make sure they are set via the `_attr_` and not directly.

File: custom_components/rpi_gpio/hub.py (hunk 16 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/hub.py:33]: fixed.

File: custom_components/rpi_gpio/binary_sensor.py (hunk 14 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/binary_sensor.py:70]: fixed.

File: custom_components/rpi_gpio/binary_sensor.py (hunk 8 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/binary_sensor.py:75]: fixed.

File: custom_components/rpi_gpio/binary_sensor.py (hunk 18 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/binary_sensor.py:76]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 15 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:92]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 9 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:98]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 19 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:107]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 22 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:120]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 21 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:127]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 23 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:134]: fixed.

File: custom_components/rpi_gpio/cover.py [hunk 24]
```diff
@@ -50,88 +53,100 @@
...
+        # self.is_closed = None
```
[the remark this answers is not in the document]
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:135]: fixed.

File: custom_components/rpi_gpio/switch.py (hunk 17 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/switch.py:76]: fixed.

File: custom_components/rpi_gpio/switch.py (hunk 10 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/switch.py:81]: fixed.

File: custom_components/rpi_gpio/switch.py (hunk 5 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/switch.py]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 23 above)
Comment (thecode) [reply custom_components/rpi_gpio/cover.py:134]: Where is it fixed?

Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:134]: Will submit in 2 hours.

**Comment by tomer-w** [comment]
Merged fixes for most of the comments.

File: custom_components/rpi_gpio/cover.py (hunk 22 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 21 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 20 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 19 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 9 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py]: fixed.

File: custom_components/rpi_gpio/hub.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tomer-w) [reply custom_components/rpi_gpio/hub.py]: fixed.

File: custom_components/rpi_gpio/binary_sensor.py (hunk 12 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/binary_sensor.py:89]: I didn't notice something happens twice. @thecode, why do you suspect that? Maybe I'm missing something.

File: custom_components/rpi_gpio/switch.py (hunk 13 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/switch.py]: I didn't notice something happens twice. @thecode, why do you suspect that? Maybe I'm missing something.

**Comment by tomer-w** [comment]
I think I handled all the CR comments other than the question about update happens twice which I didn't understand. Let me know if anything else is missing.

File: custom_components/rpi_gpio/hub.py [hunk 25]
```diff
@@ -0,0 +1,189 @@
...
+    "AS_IS": Bias.AS_IS,
+}
+DRIVE = { 
+    "OPEN_DRAIN": Drive.OPEN_DRAIN, 
+    "OPEN_SOURCE": Drive.OPEN_SOURCE, 
+    "PUSH_PULL": Drive.PUSH_PULL, 
+} 
+
+class Hub:
+
```
Comment (thecode) [inline custom_components/rpi_gpio/hub.py:32]:
```suggestion
```

File: custom_components/rpi_gpio/hub.py [hunk 26]
```diff
@@ -0,0 +1,188 @@
...
+                # rpi3,4 using 0. rpi5 using 4
+                path = f"/dev/gpiochip{d}"
+                if self.verify_gpiochip(path):
+                    self._online = True
+                    self._path = path
+                    break
+
+        if not self._online:
+            _LOGGER.error("No gpio device detected, bailing out")
+            raise IntegrationError("No gpio device detected")
```
Comment (thecode) [inline custom_components/rpi_gpio/hub.py:64]:
```suggestion
            raise HomeAssistantError("No gpio device detected")
```
`IntegrationError` is internal for HA, integrations should raise `HomeAssistantError`

File: custom_components/rpi_gpio/hub.py [hunk 27]
```diff
@@ -0,0 +1,188 @@
+from __future__ import annotations
+
+from . import DOMAIN
+
+import logging
+_LOGGER = logging.getLogger(__name__)
+
+from homeassistant.core import HomeAssistant
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP, EVENT_HOMEASSISTANT_START
+from homeassistant.exceptions import IntegrationError
```
Comment (thecode) [inline custom_components/rpi_gpio/hub.py:10]:
```suggestion
from homeassistant.exceptions import HomeAssistantError
```

File: custom_components/rpi_gpio/binary_sensor.py (hunk 12 above)
Comment (thecode) [reply custom_components/rpi_gpio/binary_sensor.py:89]: You can't notice if something update twice, only if you enable debug logging for core (which is really hard to read later) or check the DB.
What you can do is remove this line, if everything is updated correctly (for example when you change the input state it updates in HA) it means it is not necessary, from looking at Home Assistant core code this should not be needed, please verify.

File: custom_components/rpi_gpio/cover.py [hunk 28]
```diff
@@ -50,88 +53,99 @@
...
-        sleep(self._relay_time)
-        write_output(self._relay_pin, 0 if self._invert_relay else 1)
+        self._attr_is_closed = self._hub.update(self._state_port)
+        self.schedule_update_ha_state(False)
 
     def close_cover(self, **kwargs):
-        """Close the cover."""
-        if not self.is_closed:
-            self._trigger()
+        if self._attr_is_closed:
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:117]:
```suggestion
        if self.is_closed:
```
I know this is confusing, but writing should use the `_attr_` and read direct from the property. This allow HA to cache values.

File: custom_components/rpi_gpio/cover.py [hunk 29]
```diff
@@ -50,88 +53,99 @@
...
+            return
+        self._hub.turn_on(self._relay_port)
+        self._attr_is_closing = True
+        self.schedule_update_ha_state(False)
+        sleep(self._relay_time)
+        if not self._attr_is_closing:
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:123]:
```suggestion
        if not self.is_closing:
```

File: custom_components/rpi_gpio/cover.py [hunk 30]
```diff
@@ -50,88 +53,99 @@
...
+            return
+        self._hub.turn_off(self._relay_port)
+        self._attr_is_closing = False
+        self.update()
 
     def open_cover(self, **kwargs):
-        """Open the cover."""
-        if self.is_closed:
-            self._trigger()
+        if not self._attr_is_closed:
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:131]:
```suggestion
        if not self.is_closed:
```

File: custom_components/rpi_gpio/cover.py [hunk 31]
```diff
@@ -50,88 +53,99 @@
...
+            return
+        self._hub.turn_on(self._relay_port)
+        self._attr_is_opening = True
+        self.schedule_update_ha_state(False)
+        sleep(self._relay_time)
+        if not self._attr_is_opening:
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:137]:
```suggestion
        if not self.is_opening:
```

File: custom_components/rpi_gpio/cover.py [hunk 32]
```diff
@@ -50,88 +53,99 @@
...
+            # opening stopped
+            return
+        self._hub.turn_off(self._relay_port)
+        self._attr_is_opening = False
+        self.update()
+
+    def stop_cover(self, **kwargs):
+        if not (self._attr_is_closing or self._attr_is_opening):
```
Comment (thecode) [inline custom_components/rpi_gpio/cover.py:145]:
```suggestion
        if not (self.is_closing or self.is_opening):
```

**Review by thecode — CHANGES_REQUESTED** [review]
Minor last comments, please fix and we should be ready to merge this

File: custom_components/rpi_gpio/hub.py (hunk 27 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/hub.py:10]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 28 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:117]: You always learn something new. Where do you learn that from?

Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:117]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 29 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:123]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 30 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:131]: fixed.

File: custom_components/rpi_gpio/cover.py (hunk 32 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/cover.py:145]: fixed.

File: custom_components/rpi_gpio/hub.py (hunk 26 above)
Comment (tomer-w) [reply custom_components/rpi_gpio/hub.py:64]: fixed.

**Comment by tomer-w** [comment]
Latest CR comments applied.

**Comment by tomer-w** [comment]
@thecode, going back to the update twice issue, your assumption is that:
```self.is_on = self._hub.update(self._port)```
will also get HA to update this state so:
```self.schedule_update_ha_state(False)```
is not needed?

**Comment by tomer-w** [comment]
@jdeneef , tagging you for the issue about the above issue.

**Review by thecode — APPROVED** [review]

**Comment by thecode** [comment]
> I can confirm from testing that the `update` function in `switch` will not get called at all as flipping the switch does not trigger fd update. It is called only for `binary_sensor` updates when the `libgpio` sense change in the voltage.

My confusing was becasue `update` is also a method used by HA to call when it needs to update an entity and in this case there is no need to call `self.schedule_update_ha_state`, however if it is called from another source you do need to call `self.schedule_update_ha_state`.

**Comment by tomer-w** [comment]
I had the same confusion initially and I think we should rename our ```update``` to something which will make it clear it is related only to our integration.

**Comment by tomer-w** [comment]
@jdeneef, I think your checking from last week broke cover: https://github.com/jdeneef/ha_gpiod/commit/ef70c3379822f280c81ed0ad254bb9af15288017
In the cover case entity is null and this line breaks:
```
entity.is_on = True if value ^ active_low else False
```
@thecode, we can't integrate it till this is fixed.

**Comment by thecode** [comment]
I have converted the issue to a discussion, we can continue the planning on the discussion: https://github.com/thecode/ha-rpi_gpio/discussions/285

**Pull request merged**